src/sync.js

```diff
@@ -621,21 +622,30 @@ SyncApi.prototype._sync = async function(syncOptions) {
         await client.store.setSyncData(data);
     }
 
+    const syncEventData = {
+        oldSyncToken: syncToken,
+        nextSyncToken: data.next_batch,
+        catchingUp: this._catchingUp,
+    };
+
+    if (this.opts.crypto && !isCachedResponse) {
+        // tell the crypto module we're about to process a sync
+        // response
+        await this.opts.crypto.onSyncWillProcess(syncEventData);
+    }
+
     try {
-        await this._processSyncResponse(syncToken, data, isCachedResponse);
+        await this._processSyncResponse(syncEventData, data, isCachedResponse);
     } catch(e) {
         // log the exception with stack if we have it, else fall back
         // to the plain description
         console.error("Caught /sync error", e.stack || e);
     }
 
-    // emit synced events
-    const syncEventData = {
-        oldSyncToken: syncToken,
-        nextSyncToken: data.next_batch,
-        catchingUp: this._catchingUp,
-    };
```

Comment:
This got declared earlier so we can pass the same data structure into the various different hooks (lots of them need to know if this is an initial sync or not, etc). The `catchingUp` flag therefore needs to be updated once the sync is actually processed.